system/jlib/jfile.cpp

```diff
@@ -749,74 +749,94 @@ bool CFile::remove()
 
-#ifndef _WIN32 // NB: this is primary here to help track down the issue reported in HPCC-28454
-        // why did rename fail? - check that src and path of dst exists (and/or is accessible), and trace it's gid and uid, size and is_dir status.
-
-        DBGLOG("%s-doRename(%s, %s) FAILED (attempt %u)", callerName, src, dst, retry+1);
+#ifndef _WIN32 // NB: this is primarily here to help track down the issues reported in HPCC-28454 and HPCC-29812
+        /*
+         * In testing 2 [rare] issues have been seen after rename returned -1:
+         * 1) The file was renamed (dst exists and src does not).
+         *    If 'manualRenameChk' is on, we check for this and return success.
+         * 2) Neither src nor dst exist, however, after a short time, dst appears.
+         *    Assuming 'numRenameRetries'is >0, this case is covered below by looping and trying again.
+         */
 
-        struct stat statInfo;
         CDateTime modTime;
-        StringBuffer modTimeStr;
-        if (0 != stat(src, &statInfo))
-            DBGLOG("src stat[error=%d ('%s')]", errno, strerror(errno));
-        else
+        struct stat statInfo;
+        StringBuffer srcModTimeStr, dstModTimeStr;
+        int srcStatErr = stat(src, &statInfo);
+        if (0 == srcStatErr)
         {
             timetToIDateTime(&modTime, statInfo.st_mtime);
-            modTime.getString(modTimeStr);
-            DBGLOG("src stat[uid=%u, gid=%u, size=%" I64F "u, dir=%s, modtime=%s]", (unsigned)statInfo.st_uid, (unsigned)statInfo.st_gid, (offset_t)statInfo.st_size, boolToStr(S_ISDIR(statInfo.st_mode)), modTimeStr.str());
+            modTime.getString(srcModTimeStr);
         }
-
-        StringBuffer dstPath;
-        splitDirTail(dst, dstPath);
-
-        if (0 != stat(dstPath, &statInfo))
-            DBGLOG("dst-path stat[error (%d) : %s", errno, strerror(errno));
-        else
+        int dstStatErr = stat(dst, &statInfo);
+        if (0 == dstStatErr)
         {
             timetToIDateTime(&modTime, statInfo.st_mtime);
-            modTime.getString(modTimeStr.clear());
-            DBGLOG("dst-path stat[uid=%u, gid=%u, size=%" I64F "u, dir=%s, modtime=%s]", (unsigned)statInfo.st_uid, (unsigned)statInfo.st_gid, (offset_t)statInfo.st_size, boolToStr(S_ISDIR(statInfo.st_mode)), modTimeStr.str());
+            modTime.getString(dstModTimeStr);
+        }
+
+        if (0 != srcStatErr) // probably means src does now not exist
+        {
+            if (manualRenameChk) // manually validate if rename() actually succeeded
+            {
+                if (0 == dstStatErr) // dst now exists
+                {
+                    DBGLOG("%s-doRename(%s, %s) FAILED BUT DID RENAME (dst modtime=%s)", callerName, src, dst, dstModTimeStr.str());
+                    return;
+                }
+            }
         }
+        VStringBuffer errMsg("%s-doRename(%s, %s) FAILED [err=%d] ", callerName, src, dst, err);
+        errMsg.appendf("src stat:[err=%d", srcStatErr);
```

Comment:
minor comment -
if srcStatErr is not 0, errno is almost surely ENOENT, but might still be interesting to log the errno.
and same for dstStatErr.